### PR TITLE
SIMSBIOHUB-562: Export Survey Data

### PR DESCRIPTION
**Pull request by NickPhura**
## Links to Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-562

### API
New endpoint that fetches and uploads survey data to s3, and returns a pre-signed url to download it.

### APP
New dialog (under survey settings dropdown button) that allows users to execute an export. Currently restricted to project coordinators and collaborators (and system/data admins).

## Testing Notes

Pre-req: S3 credentials in `.env`

From the settings button at the top of the survey page, clicking export pops a dialogue, which lets you choose which data to export. Data is exported to a zip file in S3, and immediately downloaded to the users local machine. 

### API
Currently the export only supports survey metadata, observations (which both just make a very basic select query, and leave the response as json), and telemetry (which returns the same data as the summary telemetry page).

### APP
The dialogue is fairly basic, with no loading bars, etc, for longer exports. It also just auto-downloads the file, which is maybe not what we want if its large. Might be better to just show a link, and let the user trigger it in their browser separately.

## Note
Export button is in the dropdown on the survey page.
image

### CODE REVIEW

File: api/src/services/export-services/export-service.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      archiveStream.pipe(s3UploadStream);
+
+      // Start the S3 upload
+      const uploadPromise = uploadStreamToS3(s3UploadStream, EXPORT_ARCHIVE_MIME_TYPE, exportConfig.s3Key);
+
+      await Promise.all(
+        exportConfig.exportStrategies.map(async (exportStrategy) => {
+          const exportStrategyConfig = await exportStrategy.getExportStrategyConfig(this.connection);
+
+          defaultLog.silly({ label: 'export', message: 'exportStrategyConfig', exportStrategyConfig });
```
Comment (mauberti-bc) [inline api/src/services/export-services/export-service.ts]: Is silly intentional here?

Comment (NickPhura) [reply api/src/services/export-services/export-service.ts]: Yeah, though its questionable if its needed. I could probably remove it as it probably doesnt add any real value.

Comment (NickPhura) [reply api/src/services/export-services/export-service.ts]: Its easy enough to know what the `exportStrategyConfig` is, so logging it probably doesnt help

Comment (NickPhura) [reply api/src/services/export-services/export-service.ts]: removed

File: api/src/services/export-services/export-service.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+   *
+   * @param {ExportDataConfig} config
+   * @return {*}  {Promise<string[]>}
+   * @memberof ExportService
+   */
+  async export(exportConfig: ExportConfig): Promise<string[]> {
+    defaultLog.debug({ label: 'export', message: 'exportConfig', exportConfig });
+
+    if (exportConfig.exportStrategies.length === 0) {
+      throw new Error('No export strategies have been.');
```
Comment (mauberti-bc) [inline api/src/services/export-services/export-service.ts]: No export strategies have been defined?

Comment (NickPhura) [reply api/src/services/export-services/export-service.ts]: Oh haha, fixed

**Comment by NickPhura** [comment]
Added some unit tests

**Review by mauberti-bc — APPROVED** [review]
👍

**Pull request merged**